### PR TITLE
Linux/wayland: clipboard

### PR DESCRIPTION
**Pull request by romgrk**
Wayland clipboard support.

The clipboard operations are blocking because that's the behavior on all other platforms & libraries I have inspected (including macOS). Blocking is also better for simplicity & correctness, because async clipboard operations could introduce the possibility of out-of-order operations (e.g. `ctrl-v` then `a` might insert "a" before pasting).

This PR only implements clipboard selection, but we will also need primary selection support eventually.

### CODE REVIEW

File: crates/gpui/src/platform/linux/wayland/clipboard.rs [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+    }
+}
+
+pub(crate) fn setup_offer_read(offer: wl_data_offer::WlDataOffer) -> anyhow::Result<FileDescriptor> {
+    let pipe = Pipe::new().map_err(anyhow::Error::msg)?;
+    offer.receive(TEXT_MIME_TYPE.to_string(), unsafe { BorrowedFd::borrow_raw(pipe.write.as_raw_fd()) });
+    Ok(pipe.read)
+}
+
+pub(crate) fn read_pipe_with_timeout(mut file: FileDescriptor) -> Result<String> {
```
Comment (romgrk) [inline crates/gpui/src/platform/linux/wayland/clipboard.rs:114]: This implementation comes from wezterm, just like the `filedescriptor` crate which was made for this precise use-case.

File: crates/gpui/src/platform/linux/wayland/clipboard.rs [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+
+    pub fn receive_mime_type(&mut self, mime: &str) {
+        self.has_text = self.has_text || mime == TEXT_MIME_TYPE;
+    }
+
+    pub fn send_source(&mut self, mime_type: &str, fd: OwnedFd) {
+        if let Some(ref mut source_content) = self.source_content {
+            let mut file = FileDescriptor::new(fd);
+            let _ = file.write(source_content.as_bytes());
+        }
```
Comment (romgrk) [inline crates/gpui/src/platform/linux/wayland/clipboard.rs:44]: Ignoring the `file.write()` result, not sure if there would be a better behavior here.

File: crates/gpui/src/platform/linux/wayland/clipboard.rs [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+    }
+
+    pub fn read(&self, client: &WaylandClient) -> Option<String> {
+        let Some(ref offer) = self.offer else {
+            return None;
+        };
+        let read_pipe = setup_offer_read(offer.clone()).unwrap();
```
Comment (romgrk) [inline crates/gpui/src/platform/linux/wayland/clipboard.rs:51]: Inconsistent with above, we `.unwrap()` if the clipboard read fails. Again, not sure what's the best behavior.

**Comment by romgrk** [comment]
This is ready for review. Beyond the comments above, this isn't terribly pretty but there's going to be some refactoring needed after #8314 so I'd rather leave this as it is.

**Comment by romgrk** [comment]
Ok, this is ready for review. For now I've only implemented clipboard selection support, but eventually we'll also need primary selection support (@mikayla-maki this can be added to the roadmap issue).

File: crates/gpui/src/platform/linux/wayland/clipboard.rs [hunk 4]
```diff
@@ -0,0 +1,164 @@
...
+
+impl Clipboard {
+    pub fn new() -> Self {
+        Self {
+            offer: None,
+            offer_has_text: false,
+            offer_is_self: false,
+            source: None,
+            source_content: None,
+            self_mime: format!("pid/{}", std::process::id()),
```
Comment (romgrk) [inline crates/gpui/src/platform/linux/wayland/clipboard.rs:63]: Inspired from glfw, to support self-copy/paste while being blocking, we add this special mime type to the list of supported types. If we see it, we know the offer comes from us and we can skip the IO and just clone the clipboard source content.

File: crates/gpui/src/platform/linux/wayland/clipboard.rs [hunk 5]
```diff
@@ -0,0 +1,165 @@
...
+fn setup_offer_read(offer: &WlDataOffer) -> anyhow::Result<FileDescriptor> {
+    let pipe = Pipe::new().map_err(anyhow::Error::msg)?;
+    offer.receive(TEXT_MIME_TYPE.to_string(), unsafe {
+        BorrowedFd::borrow_raw(pipe.write.as_raw_fd())
+    });
+    Ok(pipe.read)
+}
+
+// sourced from https://github.com/wez/wezterm/blob/95581d8697f3749f84ccb1402ac94ea6582b227f/window/src/os/wayland/window.rs#L468
+fn read_pipe_with_timeout(mut file: FileDescriptor) -> Result<String> {
```
Comment (romgrk) [inline crates/gpui/src/platform/linux/wayland/clipboard.rs:128]: Slightly modified version of this function coming from wezterm, which is MIT licensed.

File: crates/gpui/src/platform/linux/wayland/clipboard.rs [hunk 6]
```diff
@@ -0,0 +1,165 @@
...
+
+    pub fn set_source(&mut self, source: WlDataSource, item: ClipboardItem) {
+        self.source = Some(source);
+        self.source_content = Some(item.text);
+    }
+
+    pub fn send_source(&mut self, mime_type: &str, fd: OwnedFd) {
+        if let Some(ref mut source_content) = self.source_content {
+            let mut file = FileDescriptor::new(fd);
+            let _ = file.write(source_content.as_bytes());
```
Comment (romgrk) [inline crates/gpui/src/platform/linux/wayland/clipboard.rs:107]: I've made both clipboard read & write fail silently rather then crashing when an IO operation fails, it seemed like the best behavior for this case.

Comment (romgrk) [reply crates/gpui/src/platform/linux/wayland/clipboard.rs]: Changed it to an error log.

**Comment by romgrk** [comment]
I have merged the calloop/main work and cleaned up the implementation a bit, this is ready for review.

**Comment by romgrk** [comment]
https://github.com/zed-industries/zed/pull/8365#issuecomment-1963365773

**Comment by flukejones** [comment]
I think the one thing that sticks out for me here is that a ctrl+c & ctrl+v on a line without selection of text pastes the line where the cursor is rather than below the line the cursor is on.

But that could be fixed later. This otherwise works very well.

**Comment by mikayla-maki** [comment]
This was a very useful exploration into how we could implement a clipboard abstraction ourselves. 

However, having looked at it, I think we'd prefer to go with a solution like alacritty does in the `copypasta` crate, using `smithay-clipboard` and `x11-clipboard`.

**Pull request closed without merging**